boxtree/tree.py

```diff
@@ -111,10 +115,111 @@ class box_flags_enum(Enum):  # noqa
 # }}}
 
 
+# {{{ tree of boxes
+
+@dataclass
+class TreeOfBoxes:
+    """A quad/octree tree of pure boxes, excluding their contents (e.g. particles).
+    It is a lightweight trees handled with :mod:`numpy`, intended for mesh
+    adaptivity. One may generate :class:`Tree` objects from a :class:`TreeOfBoxes`
+    object for FMM by distributing particles into the boxes, which often corresponds
+    to applying a quadrature rule to each leaf box.
+
+    .. attribute:: dimensions
+
+    .. attribute:: nlevels
+
+    .. attribute:: nboxes
+
+    .. attribute:: root_extent
+
+        (Scalar) extent of the root box.
+
+    .. attribute:: box_centers
+
+        dim-by-nboxes :mod:`numpy` array of the centers of the boxes.
+
+    .. attribute:: box_parent_ids
+
+        :mod:`numpy` vector of parent box ids.
+
+    .. attribute:: box_child_ids
+
+        (2**dim)-by-nboxes :mod:`numpy` array of children box ids.
+
+    .. attribute:: box_levels
+
+        :mod:`numpy` vector of box levels in non-decreasing order.
+
+    .. automethod:: __init__
+
+    .. automethod:: get_leaf_flags
+
+    .. automethod:: get_leaf_boxes
+    """
+    box_centers: npt.NDArray
+    root_extent: npt.NDArray
+    box_parent_ids: npt.NDArray
+    box_child_ids: npt.NDArray
+    box_levels: npt.NDArray
+
+    def __post_init__(self):
+        if isinstance(self.box_centers, cl.array.Array):
+            bcenters = self.box_centers.get()
+        else:
+            bcenters = self.box_centers
+        lows = bcenters[:, 0] - 0.5*self.root_extent
+        highs = lows + self.root_extent
+        self.bounding_box = [lows, highs]
+        self.dim = self.box_centers.shape[0]
+
+    # {{{ dummy interface for TreePlotter
+
+    @property
+    def dimensions(self):
+        return self.dim
+
+    def get_box_size(self, ibox):
+        lev = self.box_levels[ibox]
+        box_size = self.root_extent * 0.5**lev
+        return box_size
+
+    def get_box_extent(self, ibox):
+        box_size = self.get_box_size(ibox)
+        extent_low = self.box_centers[:, ibox] - 0.5*box_size
+        extent_high = extent_low + box_size
+        return extent_low, extent_high
+
+    # }}} End dummy interface for TreePlotter
+
+    @property
+    def nboxes(self):
+        return self.box_centers.shape[1]
+
+    @property
+    def nlevels(self):
+        # level starts from 0
+        if isinstance(self.box_levels, cl.array.Array):
+            return int(max(self.box_levels).get()) + 1
+        else:
+            return max(self.box_levels) + 1
+
+    def get_leaf_flags(self):
+        # box_id -> whether the box is leaf
+        return np.all(self.box_child_ids == 0, axis=0)
+
+    def get_leaf_boxes(self):
+        boxes = np.arange(self.nboxes)
+        return boxes[self.get_leaf_flags()]
```

Comment:
Why not make them all `@property`-ies?

Reply:
a1e535947c3b4da5382201409d5d079e6ede5a98